### ISSUE TITLE
Support `tls` option for `kn domain create`

### ISSUE THREAD

**Issue by nak3** [knative/client#1408]
### Feature request

- DomainMapping supports `spec.tls` field by https://github.com/knative/serving/pull/11250.
- I would like to request `kn domain create` supports an option to fill in the `spec.tls` field.

### Use case

- As an Knative user, I would like to create DomainMapping with custom certificate by kn command so that I can generate the DomainMapping by only one command.

### UI Example

`kn domain create` supports `--tls` option. For example, the following command:

```
$ kn domain create hello-example.default.example.org --tls=my-certs --ref=hello-example
```

creates this DomainMapping:

```
apiVersion: serving.knative.dev/v1beta1
kind: DomainMapping
metadata:
  annotations:
  name: hello-example.default.example.org
  namespace: default
spec:
  ref:
    apiVersion: serving.knative.dev/v1
    kind: Service
    name: hello-example
    namespace: default
  tls:
    secretName: my-certs
```

/kind proposal
/cc  @dsimansk 
FYI @shinigambit @julz

**Comment by rhuss** [comment]
@nak3 just to be sure, the `tls:` option has two functions: (a) enabling TLS for the domain (i.e. supporting `https://`) *and* (b) allowing to reference a secret that holds the server certificate ? There is no mode where you only enable TLS, but use e.g. self-signed, auto-generated certs ?

**Comment by nak3** [comment]
Yes, that's correct. We have to manually set a secret to the `tls:` field (b) and it prints `https` schema(a).
There is no any other supported usage/mode.

**Comment by julz** [comment]
I think domainmapping also supports autotls if it's enabled in the cluster, but you don't have to do anything for this it just works

**Comment by rhuss** [comment]
My question actually is, would `--tls` always need a value or is it possible to just have

```
apiVersion: serving.knative.dev/v1beta1
kind: DomainMapping
metadata:
  annotations:
  name: hello-example.default.example.org
  namespace: default
spec:
  ref:
    apiVersion: serving.knative.dev/v1
    kind: Service
    name: hello-example
    namespace: default
  tls: {}
```

**Comment by nak3** [comment]
Thank you for the following up! Let me rephrase it :sweat_smile:   

- The function/usage of `spec.tls` is only to enable TLS and to display `https` schema by setting a secret manually.
- DomainMapping also supports autoTLS but we enable it on `config-network` configmap, it does not use `spec.tls` in DomainMapping. 

Sorry, my answer "There is no any other supported usage/mode." was bad... It sounds like DomainMapping does not support autoTLS :disappointed:

**Comment by nak3** [comment]
> My question actually is, would --tls always need a value or is it possible to just have

`spec.tls` always needs a value (`secretName`) and so It does not allow empty `tls: {}`. I think it is by design (because no usecase for the empty `tls: {}`) but @julz or @shinigambit know more.

**Comment by julz** [comment]
yeah - you only need to specify `tls` if you want to specify your own cert. Otherwise you either (a) dont have tls on in the cluster -> http (b) do have autotls on in the cluster -> tls, with generated certs (c) do have tls on in the cluster but want to opt out a specific ksvc/domainmapping, in which case you use the `networking.knative.dev/disableAutoTLS` annotation -> http

**Comment by rhuss** [comment]
Thanks for the clarification, the only thing I don't fully understand case (c ). If I understand, this case applies when autotls is enabled in the global knative configuration, but for a specific service you want to have HTTP , so you add the annotation to the service. Or can this annotation also be used on the DomainMapping ?

I.e. for this configuration 

```yaml
apiVersion: serving.knative.dev/v1beta1
kind: DomainMapping
metadata:
  annotations:
    networking.knative.dev/disableAutoTLS: true
  name: hello-example.default.example.org
  namespace: default
spec:
  ref:
    apiVersion: serving.knative.dev/v1
    kind: Service
    name: hello-example
    namespace: default
  tls: 
    secretName: my-certs
```

would allow for a custom cert for this domain, even when autoTLS is enabled? What would happen if the annotation would be missing in this case? Would the custom cert still "overrule" the autotls auto-generated certs ? So that the annotation is only needed if no `tls:` section is given ?

**Comment by julz** [comment]
If a TLS section is explicitly provided it's always used, regardless whether autotls is enabled in the cluster or not (so the disableAutoTLS annotation is ignored in this case, 'cos no autotls is happening). Otherwise (no explicit TLS section) stuff works just like Ksvcs (autotls if enabled on the cluster, unless disabled by disableAutoTLS annotation). Here's the code to check: https://github.com/knative/serving/blob/main/pkg/reconciler/domainmapping/reconciler.go#L202-L215

**Comment by julz** [comment]
To put it another way, the TLS section is independent of autotls. You can specify explicit TLS certs *or* you can fall back to the cluster-wide autoTLS, if enabled in the cluster (and you can opt out of that autoTLS, as normal, with the disableAutoTLS annotation - but that doesn't affect explicit TLS because explicit TLS isn't "auto" :))
